applications/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -1025,6 +1046,176 @@ static void send_enqueued_report(struct subscriber *sub)
 	}
 }
 
+static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info *rep, uint8_t err);
+
+static int send_hid_out_report(struct bt_hogp *hogp, const uint8_t *data, size_t size)
+{
+	uint8_t report_id = data[0];
+	struct bt_hogp_rep_info *out_rep = bt_hogp_rep_find(hogp,
+							    BT_HIDS_REPORT_TYPE_OUTPUT,
+							    report_id);
+
+	if (!out_rep) {
+		/* HID output report with given ID is not supported by the peripheral. */
+		return 0;
+	}
+
+	int err = bt_hogp_rep_write_wo_rsp(hogp, out_rep,
+					   &data[1], size - 1,
+					   hogp_out_rep_write_cb);
+
+	if (err == -ENOTCONN) {
+		LOG_INF("Cannot forward, peer disconnected");
+		err = 0;
+	}
+
+	return err;
+}
+
+static void hogp_out_rep_write_cb(struct bt_hogp *hogp, struct bt_hogp_rep_info *rep, uint8_t err)
+{
+	/* Make sure that the handler will not preempt system workqueue
+	 * and it will not be preempted by system workqueue.
+	 */
+	__ASSERT_NO_MSG(!k_is_in_isr());
+	__ASSERT_NO_MSG(!k_is_preempt_thread());
+
+	if (err) {
+		LOG_WRN("Failed to forward HID out report (err %" PRIu8 ")", err);
+	}
+
+	struct hids_peripheral *per = CONTAINER_OF(hogp, struct hids_peripheral, hogp);
+
+	/* Send update if it was queued. */
+	uint8_t sent_report_id = bt_hogp_rep_id(rep);
+	sys_snode_t *cur_node;
+	struct enqueued_out_report *found_report = NULL;
+
+	SYS_SLIST_FOR_EACH_NODE(&per->enqueued_out_reports, cur_node) {
+		struct enqueued_out_report *cur_report = CONTAINER_OF(cur_node,
+								      struct enqueued_out_report,
+								      node);
+
+		uint8_t cur_report_id = cur_report->data[0];
+
+		if (cur_report_id == sent_report_id) {
+			found_report = cur_report;
+			break;
+		}
+	}
+
+	if (!found_report) {
+		return;
+	}
+
+	int send_err = send_hid_out_report(hogp, found_report->data, found_report->data_size);
+
+	if (send_err) {
+		LOG_ERR("Cannot forward HID report (err: %d)", send_err);
+	}
+
+	sys_slist_find_and_remove(&per->enqueued_out_reports, &found_report->node);
+	k_free(found_report->data);
+	k_free(found_report);
+}
+
+static void append_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
+{
+	struct enqueued_out_report *item = k_malloc(sizeof(*item));
+
+	if (!item) {
+		LOG_ERR("Failed to allocate memory to forward HID output report");
+		return;
+	}
+
+	item->data_size = size;
+	item->data = k_malloc(size);
+
+	if (!item->data) {
+		k_free(item);
+		LOG_ERR("Failed to allocate memory to forward HID output report");
+		return;
+	}
+
+	memcpy(item->data, data, size);
+	sys_slist_append(list, &item->node);
+};
+
+static void enqueue_hid_out_report(sys_slist_t *list, const uint8_t *data, size_t size)
```

Comment:
I am just wondering. We have report enqueueing system already for in_reports... is there a way to have it reused for out_reports (I mean functions not the actual queue) instead of writing duplicates. Of there are enough changes for it to be impractical?

Reply:
This is quite different. The HID input reports can be enqueued as hid_report_event since they are going to be submitted and handled by the Event Manager. We cannot "catch" the hid_out_report event in similar way (still we may need to temporarily store the data related to the report).

It seems like the only common thing is the fact that we use linked list in both cases.